google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/apis/CloudLibraryProjectStateTest.java

```diff
@@ -0,0 +1,242 @@
+import com.intellij.openapi.util.Computable;
+import com.intellij.testFramework.fixtures.IdeaProjectTestFixture;
+import com.intellij.util.xml.DomUtil;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import org.jdom.JDOMException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.idea.maven.dom.MavenDomUtil;
+import org.jetbrains.idea.maven.dom.model.MavenDomProjectModel;
+import org.jetbrains.idea.maven.model.MavenId;
+import org.jetbrains.idea.maven.project.MavenProject;
+import org.jetbrains.idea.maven.project.MavenProjectsManager;
+import org.jetbrains.idea.maven.server.MavenServerManager;
+import org.jetbrains.idea.maven.wizards.MavenModuleBuilder;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+
+/** Tests for {@link CloudLibraryProjectState}. */
+public class CloudLibraryProjectStateTest {
+
+  @Rule public final CloudToolsRule cloudToolsRule = new CloudToolsRule(this);
+  @TestFixture private IdeaProjectTestFixture testFixture;
+
+  @Mock @TestService CloudLibrariesService librariesService;
+
+  @TestDirectory(name = "root")
+  private File root;
+
+  private static final TestCloudLibraryClientMavenCoordinates JAVA_CLIENT_MAVEN_COORDS =
+      TestCloudLibraryClientMavenCoordinates.create("java", "client", "1.0.0");
+  private static final TestCloudLibraryClient JAVA_CLIENT =
+      TestCloudLibraryClient.create(
+          "Client", "java", "API Ref", "alpha", "Source", "Lang Level", JAVA_CLIENT_MAVEN_COORDS);
+  private static final TestCloudLibrary LIBRARY =
+      TestCloudLibrary.create(
+          "Library",
+          "ID",
+          "service",
+          "Docs Link",
+          "Description",
+          "Icon Link 1",
+          ImmutableList.of(JAVA_CLIENT));
+
+  private CloudLibraryProjectState state;
+  private MavenModuleBuilder moduleBuilder;
+  private MavenId id;
+
+  @Before
+  public void setUp() {
+    state = CloudLibraryProjectState.getInstance(testFixture.getProject());
+    moduleBuilder = new MavenModuleBuilder();
+    id = new MavenId("org.foo", "module", "1.0");
+
+    setModuleNameAndRoot(testFixture.getProject().getBasePath());
+  }
+
+  @After
+  public void tearDown() {
+    MavenServerManager.getInstance().shutdown(true);
+  }
+
+  @Test
+  public void managedLibraries_withNoDependenies_isEmptyOptional() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              Module module = createNewModule(id);
+
+              assertThat(state.getManagedLibraries(module)).isEmpty();
+            });
+  }
+
+  @Test
+  public void managedLibraries_withOnlyUnmanagedDependencies_isEmptyOptional() {
+    when(librariesService.getCloudLibraries())
+        .thenReturn(ImmutableList.of(LIBRARY.toCloudLibrary()));
+
+    ApplicationManager.getApplication()
+        .invokeAndWait(
```

Comment:
Interesting, what is the reason for having `invokeAndWait` here, mostly because of `writeDependenciesToPom` or sync with `runReadAction` calls in `CloudLibraryProjectState`?

Reply:
I believe its because the write action needs to be executed from the EDT